File: cpp/questions/trap.cpp
#include <stack>
#include <vector>

using namespace std;

class Solution {
public:
    int trap(vector<int>& height) {
        std::stack<int> st;
        st.push(-1);
        int ans = 0;
        for (int i=0; i<height.size(); ++i) {
            int ch = height[i];
            int sp = i - 1;
            while (st.top()!=-1 && ch>=height[st.top()]) {
                int ct = st.top();
                st.pop();
                ans += st.top()==-1 ? (i-ct-1)*(height[ct]-ch) : (sp-st.top())*(ch-height[ct]);
                sp = st.top();
            }
            st.push(i);
        }
        return ans;
    }
};

File: rust/questions/trap.rs
impl Solution {
    pub fn trap(height: Vec<i32>) -> i32 {
        let hl = height.len();
        let mut st = Vec::new();
        let mut ans = 0;
        for (i, &h) in height.iter().enumerate() {
            while st.len() > 0 && h >= height[*st.last().unwrap()] {
                let bh = height[st.pop().unwrap()];
                if st.len() == 0 {
                    break;
                }
                ans += (h.min(height[st[st.len() - 1]]) - bh) * ((i - st[st.len() - 1]) as i32 - 1);
            }
            st.push(i);
        }
        ans
    }
}
